### PR TITLE
[SPARK-11564][SQL] Dataset Java API audit

### PR DESCRIPTION
**Pull request by rxin**
A few changes:
1. Removed fold, since it can be confusing for distributed collections.
2. Created specific interfaces for each Dataset function (e.g. MapFunction, ReduceFunction, MapPartitionsFunction)
3. Added more documentation and test cases.

The other thing I'm considering doing is to have a "collector" interface for FlatMapFunction and MapPartitionsFunction, similar to MapReduce's map function.

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan @marmbrus

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -1478,18 +1478,54 @@ class DataFrame private[sql](
 
   /**
    * Returns the first `n` rows in the [[DataFrame]].
+   *
+   * Running take requires moving data into the application's driver process, and doing so on a
+   * very large dataset can crash the driver process with OutOfMemoryError.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: and doing so with a very large `n`?  as phrased now it seems more dangerous than it is.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 2]
```diff
@@ -1478,18 +1478,54 @@ class DataFrame private[sql](
...
+  /**
+   * Returns a Java list that contains all of [[Row]]s in this [[DataFrame]].
+   *
+   * Running collect requires moving all the data into the application's driver process, and
+   * doing so on a very large dataset can crash the driver process with OutOfMemoryError.
+   *
+   * @group action
+   * @since 1.3.0
+   */
+  def collectAsList(): java.util.List[Row] = withCallback("collectAsList", this) { _ =>
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: oh hmm, not in the PR, but we should add `withCallback` to dataset too.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 3]
```diff
@@ -1478,18 +1478,54 @@ class DataFrame private[sql](
...
+   */
+  def takeAsList(n: Int): java.util.List[Row] = java.util.Arrays.asList(take(n) : _*)
+
+  /**
    * Returns an array that contains all of [[Row]]s in this [[DataFrame]].
+   *
+   * Running collect requires moving all the data into the application's driver process, and
+   * doing so on a very large dataset can crash the driver process with OutOfMemoryError.
+   *
+   * For Java API, use [[collectAsList]].
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: `which will return the correct type without casting`?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: oh wait, why?  The array variance problems don't exist for dataframe.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 4]
```diff
@@ -166,8 +171,7 @@ class Dataset[T] private[sql](
    * Returns a new [[Dataset]] that only contains elements where `func` returns `true`.
    * @since 1.6.0
    */
-  def filter(func: JFunction[T, java.lang.Boolean]): Dataset[T] =
-    filter(t => func.call(t).booleanValue())
+  def filter(func: FilterFunction[T]): Dataset[T] = filter(t => func.call(t).booleanValue())
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Get rid of `booleanValue`?  I think this is boxing.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 3 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: No it doesn't' exist, but still good to use a java list.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 4 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: good catch

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 5]
```diff
@@ -1478,18 +1478,54 @@ class DataFrame private[sql](
...
   /**
+   * Returns the first `n` rows in the [[DataFrame]] as a list.
+   *
+   * Running take requires moving data into the application's driver process, and doing so with
+   * a very large `n` can crash the driver process with OutOfMemoryError.
+   *
+   * @group action
+   * @since 1.6.0
+   */
+  def takeAsList(n: Int): java.util.List[Row] = java.util.Arrays.asList(take(n) : _*)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: In other places we usually `import scala.collection.JavaConverters._` and use `asJava` to do the conversion, should we follow that?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: I'm not sure if we need the `xxxAsList` methods. The `DataFrame` is typed with `Row` and the return type of `collect` or `take` is `Array[Row]` which works well at java side.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 6]
```diff
@@ -1478,18 +1478,54 @@ class DataFrame private[sql](
...
+   * doing so on a very large dataset can crash the driver process with OutOfMemoryError.
+   *
+   * @group action
+   * @since 1.3.0
+   */
+  def collectAsList(): java.util.List[Row] = withCallback("collectAsList", this) { _ =>
+    withNewExecutionId {
+      java.util.Arrays.asList(rdd.collect() : _*)
+    }
+  }
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: how about:
```
withCallback("collectAsList", this) { df =>
  df.collect(needCallback = false).asJava
}
```

**Comment by cloud-fan** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala (hunk 5 above)
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: take returns an array not a seq though

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 7]
```diff
@@ -1478,18 +1478,54 @@ class DataFrame private[sql](
...
+   * @group action
+   * @since 1.6.0
+   */
+  def takeAsList(n: Int): java.util.List[Row] = java.util.Arrays.asList(take(n) : _*)
+
+  /**
    * Returns an array that contains all of [[Row]]s in this [[DataFrame]].
+   *
+   * Running take requires moving data into the application's driver process, and doing so with
+   * a very large `n` can crash the driver process with OutOfMemoryError.
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: damn fixed the wrong

**Pull request closed without merging**